lib/client/client_test.go

```diff
@@ -182,6 +183,46 @@ func (s *ClientTestSuite) TestProxyConnection(c *check.C) {
 	}
 }
 
+// TestListenAndForwardCancel verifies that the "Accept" in listenAndForward
+// unblocks when canceled.
+func (s *ClientTestSuite) TestListenAndForwardCancel(c *check.C) {
+	client := &NodeClient{
+		Client: &ssh.Client{
+			Conn: &fakeSSHConn{},
+		},
+	}
+
+	// Create a new cancelable listener.
+	ctx, cancel := context.WithCancel(context.Background())
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	c.Assert(err, check.IsNil)
+
+	// Start listenAndForward using a context to find out when "Accept" has
+	// unblocked.
+	blockContext, blockCancel := context.WithCancel(context.Background())
+	go func() {
+		client.listenAndForward(ctx, ln, "")
+		blockCancel()
+	}()
+
+	// At this point, "Accept" should still be blocking.
+	select {
+	case <-blockContext.Done():
+		c.Fatalf("Failed because Accept was unblocked.")
+	case <-time.After(250 * time.Millisecond):
+	}
+
+	// Cancel "Accept" to unblock it.
+	cancel()
+
+	// Verify that "Accept" is no longer blocking.
+	select {
+	case <-blockContext.Done():
+	case <-time.After(250 * time.Millisecond):
```

Comment:
Here's an example to outline my point:
```go
func (s *ClientTestSuite) TestListenAndForwardCancel(c *check.C) {
	client := &NodeClient{
		Client: &ssh.Client{
			Conn: &fakeSSHConn{},
		},
	}

	startedC := make(chan struct{}) // First anchor for when the listener starts accepting
	doneC := make(chan struct{})    // Second anchor for when the client stops the listen loop
	// Create a new cancelable listener.
	ctx, cancel := context.WithCancel(context.Background())
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	c.Assert(err, check.IsNil)
	ln = wrapListener(ln, startedC)

	// Start listenAndForward using a context to find out when "Accept" has
	// unblocked.
	blockContext, blockCancel := context.WithCancel(context.Background())
	go func() {
		client.listenAndForward(ctx, ln, "")
		blockCancel()
		close(doneC)
	}()

	<-startedC
	// It is safe to assume the listener is accepting at this point
	// so we merely test the context expiration

	// At this point, "Accept" should still be blocking.
	select {
	case <-blockContext.Done():
		c.Fatalf("Failed because Accept was unblocked.")
	default:
	}

	// Cancel "Accept" to unblock it.
	cancel()
	<-doneC
	// I think it's fine if this blocks.
	// This will timeout in automatic tests and can be controlled on command line during manual runs

	// Verify that "Accept" is no longer blocking.
	select {
	case <-blockContext.Done():
	default:
		c.Fatalf("Accept failed to unblock.")
	}
}
// and the supporting code
func wrapListener(listener net.Listener, startedC chan struct{}) wrappedListener {
	return wrappedListener{
		Listener: listener,
		startedC: startedC,
	}
}

func (r wrappedListener) Accept() (net.Conn, error) {
	close(r.startedC)
	return r.Listener.Accept()
}

type wrappedListener struct {
	net.Listener
	startedC chan struct{}
}
```